File: backend/app/models/notification.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base


class Notification(Base):
    """Bot orqali foydalanuvchiga yuborilgan har bir xabarning nusxasi —
    ilovadagi "Bildirishnomalar" sahifasi shundan o'qiydi."""

    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    kind: Mapped[str] = mapped_column(String(32))  # order_status | broadcast
    title: Mapped[str] = mapped_column(String(255))
    body: Mapped[str] = mapped_column(Text)
    image_url: Mapped[str | None] = mapped_column(String(512))
    order_id: Mapped[int | None] = mapped_column(
        ForeignKey("orders.id", ondelete="SET NULL"), index=True
    )
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), index=True
    )

File: mijoz_app/lib/models/notification.dart
class AppNotification {
  AppNotification({
    required this.id,
    required this.kind,
    required this.title,
    required this.body,
    this.imageUrl,
    this.orderId,
    required this.isRead,
    required this.createdAt,
  });

  final int id;
  final String kind; // order_status | broadcast
  final String title;
  final String body;
  final String? imageUrl;
  final int? orderId;
  final bool isRead;
  final String createdAt;

  factory AppNotification.fromJson(Map<String, dynamic> j) => AppNotification(
        id: j['id'] as int,
        kind: (j['kind'] ?? 'order_status') as String,
        title: (j['title'] ?? '') as String,
        body: (j['body'] ?? '') as String,
        imageUrl: j['image_url'] as String?,
        orderId: j['order_id'] as int?,
        isRead: (j['is_read'] ?? false) as bool,
        createdAt: (j['created_at'] ?? '') as String,
      );

  static List<AppNotification> listFrom(dynamic json) =>
      (json as List).map((e) => AppNotification.fromJson(e as Map<String, dynamic>)).toList();
}
